src/app.tsx

```diff
@@ -0,0 +1,779 @@
+      dialog: null
+    };
+    this.save();
+  }
+
+  showImportExport() {
+    this.setDialog(<ImportExport
+      stacks={this.state.stacks}
+      import={this.import}
+      cancel={this.cancel}
+    />);
+  }
+
+  stackPopped(name: string, returnToBottom: boolean) {
+    this.setState((prevState) => {
+      let state = prevState;
+      let events = this.state.stacks[name.toLowerCase() + "Events"] as any;
+      console.log("events", events);
+      let event = events.stack.shift();
+      let action = "removed from game";
+      if (returnToBottom) {
+        events.stack.push(event);
+        action = "returned to bottom";
+      }
+      events.history.push({
+        event: event,
+        action: action,
+      });
+      return state;
+    }, this.save);
+    this.cancel();
+  }
+
+  stackDrawn(name: String, cards: CardStack, cardNo: number): void {
+    this.setState((prevState: Readonly<AppState>) => {
+      let state = prevState;
+      removeFromArray(cards.stack, cardNo);
+      cards.list.push(cardNo);
+      cards.list.sort();
+      cards.history.push({
+        action: "drawn",
+        card: cardNo,
+      });
+
+      return state;
+    }, this.save);
+    this.cancel();
+  }
+
+  import(text: string) {
+    try {
+      let stacks = App.initializeStacks(JSON.parse(text));
+      this.setState({
+        stacks: stacks
+      }, this.save);
+    } catch (e) {
+      //TODO :(
+      alert(e.message)
+    }
+    this.cancel();
+  }
+
+  addCardsAndCloseDialog(cardType: string, cardIdsToAdd: Array<number>) {
+    this.addCards(cardType, cardIdsToAdd);
+    this.cancel();
+  }
+
+  addCards(cardType: string, cardIdsToAdd: Array<number>) {
+    if (!cardIdsToAdd || cardIdsToAdd.length === 0) {
+      return;
+    }
+
+    let simpleListMappings: { [cardType: string]: CardStack } = {
+      "Item Designs": this.state.stacks.itemDesigns,
+      "Single Items": this.state.stacks.singleItems,
+    };
+    if (simpleListMappings[cardType]) {
+      this.setState((prevState) => {
+        let state = prevState;
+        let list = simpleListMappings[cardType].list.concat(cardIdsToAdd);
+        list.sort();
+        if (cardType !== "Single Items") {
+          // remove duplicates, but not on "Single Items"
+          list = list.filter((el, idx, arr) => el !== arr[idx - 1]);
+        }
+        simpleListMappings[cardType].list = list;
+        simpleListMappings[cardType].history.push({
+          action: "added cards",
+          cards: cardIdsToAdd,
+        });
+
+        return state;
+      }, this.save);
+    } else {
+      // assuming "City Event" or "Road Event"
+      const eventStackName = cardType.split(" ")[0].toLowerCase() + "Events";
+      const stack = this.state.stacks[eventStackName];
+      if (!stack || typeof stack === "number") {
+        throw "Unknown name for addCards: " + cardType;
+      }
+
+      const notAlreadyContainedCards = unique(cardIdsToAdd).filter((it) => !stack.stack.includes(it));
```

Comment:
Is this the only change? Moving from index.tsx to app.tsx AND doing code changes was a good way to hide the actual few lines of changes ;)

Reply:
I agree, I should have made two separate pull requests. Initially the fragment change was also part of this on  :)

Reply:
And yes, this is what changed.